### PR TITLE
Feature - image model #47

### LINKED ISSUE #47: Define the Icon / Image model
If we look at the designs for On-Hub notifications, there is two uses for images in the notifications:

- One is the icon that represents the **sender** (see [this comment]((https://github.com/WordPress/wp-notify/issues/41#issuecomment-773525936 )) in the Sender model discussion)
- The other is an image related to the message of the notification

Maybe both the Sender model and the Notification model would depend on a third model: the Image model?

Here are the data it should contain:

- **source**: string. An URL or data URL (see details below)
- **alternative_text**: string. Description of the image for accessibility purpose.

On the front-end, the `` tag should be able to handle all the use cases:

- An **external URL** will be resolved in a same way by its [src attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-src) whether the format of the image is PNG, JPEG or SVG.
- This `src` attribute can as well accept a [data URL](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs) to directly embed base64 encoded raster images or SVG inside the notification itself (saves an HTTP request, per notification!)

### PR DESCRIPTION
**Pull request by danbilauca**
## Goal

This PR implements the `WPNotify_BaseImage` model. 

Fixes #47

### CODE REVIEW

File: includes/image/BaseImage.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	 * @return WPNotify_JsonUnserializable
+	 */
+	public static function json_unserialize( $json ) {
+
+		$data = json_decode( $json, true );
+
+		$source = ! empty( $data['source'] ) ? $data['source'] : '';
+		$alt    = ! empty( $data['alt'] ) ? $data['alt'] : '';
+
+		return new static( $source, $alt );
```
Comment (raaaahman) [inline includes/image/BaseImage.php]: I notice that sometimes we use `new self( ...$args );` and sometimes we use ` new static( ...$args );`. Which one better suits our needs in your opinion?

File: tests/phpunit/tests/base-image.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
+<?php
+
+class Test_WPNotify_BaseImage extends WPNotify_TestCase {
+
+	/**
+	 * @param array  $image_data
+	 * @param string $expected_json
+	 *
+	 * @dataProvider data_provider_images
```
Comment (raaaahman) [inline tests/phpunit/tests/base-image.php:9]: Good idea to use the same data provider for both tests! :+1:

File: tests/phpunit/tests/base-image.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+		list( $source, $alt ) = array_values( $image_data );
+
+		$instance = new WPNotify_BaseImage( $source, $alt );
+
+		$this->assertEquals( $instance->get_source(), $test_instance->get_source() );
+		$this->assertEquals( $instance->get_alt(), $test_instance->get_alt() );
+	}
+
+	public function data_provider_images() {
```
Comment (raaaahman) [inline tests/phpunit/tests/base-image.php:42]: Your four cases provide a good test coverage for our code. Well done!

As a side note, you can name each individual case and PHPUnit will use these names in its failure reports. For example:

```php
       'With image source and alternative text' => array(
		array(
			'source' => 'img-source1',
			'alt'    => 'img-alt1',
		),
		'{"source":"img-source1","alt":"img-alt1"}',
	),

```

File: includes/image/Image.php [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<?php
+
+interface WPNotify_Image extends JsonSerializable {
```
Comment (raaaahman) [inline includes/image/Image.php]: This class also validates the `WPNotify_JsonUnserializable` interface.

File: includes/senders/BaseSender.php [hunk 5]
```diff
@@ -36,16 +44,26 @@ public static function json_unserialize( $json ) {
 
 		$data = json_decode( $json, true );
 
-		$name = ! empty( $data['name'] ) ? $data['name'] : '';
+		$name  = ! empty( $data['name'] ) ? $data['name'] : '';
+		$image = ! empty( $data['image'] ) ? WPNotify_BaseImage::json_unserialize( json_encode( $data['image'] ) ) : null;
```
Comment (raaaahman) [inline includes/senders/BaseSender.php]: What happens if we set the `$associative` parameter of `json_decode()` to false? Do we have our properties correctly instantiated as object? 

If this doesn't work, could we use `$depth` parameter of the `json_decode()` function to only decode the first level objects? We may able able to have our image as a JSON string and then not have to encode it again.

https://www.php.net/manual/fr/function.json-decode.php

File: tests/phpunit/tests/base-image.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+				),
+				'{"source":"img-source1","alt":"img-alt1"}',
+			),
+
+			array(
+				array(
+					'source' => 'img-source2',
+					'alt'    => '',
+				),
+				'{"source":"img-source2","alt":""}',
```
Comment (raaaahman) [inline tests/phpunit/tests/base-image.php]: Do you think we should encode empty properties in our JSON strings? Maybe we could only encode the non empty instead.

For example, the expected JSON here would be  `'{"source":"img-source2"}`.

File: tests/phpunit/tests/base-sender.php [hunk 7]
```diff
@@ -3,38 +3,75 @@
...
-		$senderInstance = new WPNotify_BaseSender( $sender_name );
-		$encoded        = json_encode( $senderInstance );
-		$this->assertEquals( $expected_json, $encoded );
+		$sender_instance = null;
+
+		if ( is_string( $sender ) ) {
+			$sender_instance = new WPNotify_BaseSender( $sender );
+		} else if ( is_array( $sender ) ) {
+			$sender_instance = new WPNotify_BaseSender( $sender['name'], $sender['image'] );
+		}
```
Comment (raaaahman) [inline tests/phpunit/tests/base-sender.php]: As a rule of thumb, I'd rather not have conditional logic in our tests: the less we have to think about what the test verifies, the quicker we could read them.

Maybe you could instantiate these objects in the `@dataProvider` instead. Or maybe you could use two different tests with no provider.

File: tests/phpunit/tests/base-notification.php [hunk 8]
```diff
@@ -2,7 +2,7 @@
 
 class Test_WPNotify_BaseNotification extends WPNotify_TestCase {
 
-	const JSON_SERIALIZED = '{"WPNotify_RecipientCollection":[],"WPNotify_BaseMessage":"Message","WPNotify_BaseSender":{"name":"Name 1"}}';
+	const JSON_SERIALIZED = '{"WPNotify_RecipientCollection":[],"WPNotify_BaseMessage":"Message","WPNotify_BaseSender":{"name":"Name 1","image":null}}';
```
Comment (raaaahman) [inline tests/phpunit/tests/base-notification.php]: Same comment than for the `Test_WPNotify_BaseImage::data_provider_images()`: maybe we don't want the empty `"image"` field to be printed in our JSON string. :thinking:

File: includes/image/BaseImage.php (hunk 1 above)
Comment (danbilauca) [reply includes/image/BaseImage.php]: I'd say `static` even though I could see all over the code `self`. I might have missed this one.

File: tests/phpunit/tests/base-notification.php (hunk 8 above)
Comment (danbilauca) [reply tests/phpunit/tests/base-notification.php]: fixed

File: tests/phpunit/tests/base-sender.php (hunk 7 above)
Comment (danbilauca) [reply tests/phpunit/tests/base-sender.php]: fixed

File: tests/phpunit/tests/base-image.php (hunk 6 above)
Comment (danbilauca) [reply tests/phpunit/tests/base-image.php]: fixed

File: includes/senders/BaseSender.php (hunk 5 above)
Comment (danbilauca) [reply includes/senders/BaseSender.php]: I've changed the way the image is initialised. I've used the same logic from `WPNotify_BaseNotification` instantiating the image from reflection.

File: includes/image/Image.php (hunk 4 above)
Comment (danbilauca) [reply includes/image/Image.php]: fixed

File: tests/phpunit/tests/base-image.php (hunk 3 above)
Comment (danbilauca) [reply tests/phpunit/tests/base-image.php:42]: fixed

File: includes/image/BaseImage.php (hunk 1 above)
Comment (danbilauca) [reply includes/image/BaseImage.php]: fixed

File: includes/image/BaseImage.php [hunk 9]
```diff
@@ -1,6 +1,6 @@
 <?php
 
-class WPNotify_BaseImage implements WPNotify_Image, WPNotify_JsonUnserializable {
```
Comment (raaaahman) [inline includes/image/BaseImage.php]: I didn't see the `WPNotify_JsonUnserializable` interface use here, sorry. Anyway, it is better how it is declared right now.

File: includes/senders/BaseSender.php [hunk 10]
```diff
@@ -27,10 +27,16 @@ public function __construct( $name, $image = null ) {
...
+
+		$data = array(
+			'name' => $this->name,
 		);
+
+		if ( $this->image ) {
+			$data[ get_class( $this->image ) ] = $this->image;
+		}
+
+		return array_filter( $data );
```
Comment (raaaahman) [inline includes/senders/BaseSender.php]: I am not sure what the `array_filter` function is supposed to do without passing it a callback function. Would you explain its usage here?

File: tests/phpunit/tests/base-sender.php [hunk 11]
```diff
@@ -3,74 +3,67 @@
...
 		/**
 		 * If an Image has been sent
 		 */
 		if ( $testee->get_image() ) {
-			$this->assertInstanceOf( 'WPNotify_BaseImage', $testee->get_image() );
-			$this->assertEquals( $sender_instance->get_image()->get_source(), $testee->get_image()->get_source() );
-			$this->assertEquals( $sender_instance->get_image()->get_alt(), $testee->get_image()->get_alt() );
+			$this->assertInstanceOf( get_class( $sender->get_image() ), $testee->get_image() );
+			$this->assertEquals( $sender->get_image()->get_source(), $testee->get_image()->get_source() );
+			$this->assertEquals( $sender->get_image()->get_alt(), $testee->get_image()->get_alt() );
```
Comment (raaaahman) [inline tests/phpunit/tests/base-sender.php]: If line 38 passes, is there any way for lines 45 to 46 to fail?

I mean if `$sender->get_image() === $testee->get_image()`, then their classes, `source` property and `alt` property probably match as well.

But  `$sender->get_image() === $testee->get_image()` imply that the two `image` properties refer to the same instance... Is this something we intended here?

File: tests/phpunit/tests/base-sender.php [hunk 12]
```diff
@@ -3,74 +3,67 @@
...
-			array(
+			'sender with image' => array(
 				array(
-					'name'  => 'Name 2',
-					'image' => new WPNotify_BaseImage( 'img-source', 'img-alt' ),
+					'name'               => 'Name 2',
+					'WPNotify_BaseImage' => new WPNotify_BaseImage( 'img-source', 'img-alt' ),
 				),
-				'{"name":"Name 2","image":{"source":"img-source","alt":"img-alt"}}',
+				'{"name":"Name 2","WPNotify_BaseImage":{"source":"img-source","alt":"img-alt"}}',
```
Comment (raaaahman) [inline tests/phpunit/tests/base-sender.php]: The Json serialization was meant to be used to communicate between the back-end and front-end parts of the feature.

I am not sure what the front-end part would do with a PHP class name. Wouldn't this be bothersome to have to change the data field names on the front-end if we ever want to add a new PHP class for images?

**Review by raaaahman — COMMENTED** [review]
Nice to have added explicit names to the test cases in the data providers.

I do not really understand the use of `ReflectionClass` here, is that because you wanted to pass parameter to the constructor as an array?

File: tests/phpunit/tests/base-sender.php (hunk 11 above)
Comment (raaaahman) [reply tests/phpunit/tests/base-sender.php]: PHPUnit does not perform a deep comparison between the two `WPNotify_BaseImage` objects, so we need to compare their individual properties.

File: tests/phpunit/tests/base-sender.php (hunk 12 above)
Comment (raaaahman) [reply tests/phpunit/tests/base-sender.php]: This is already the way `WPNotify_BaseNotifications` are instantiated, it's better to keep our code consistent.

**Review by raaaahman — APPROVED** [review]

**Pull request merged**